test/FunctionalTests/Client/StreamingTests.cs

```diff
@@ -996,5 +997,90 @@ async Task WaitForAllStreams(IAsyncStreamReader<DataMessage> requestStream, ISer
+                    }
+                }
+                catch (IOException ex) when (ex.Message == "The client reset the request stream.")
+                {
+                    // The server receives END_STREAM = true and then RST_STREAM. There is a race between
+                    // gRPC and Kestrel over whether gRPC finishes using PipeReader first, or Kestrel reads
+                    // the RST_STREAM and makes PipeReader throw.
+                }
+                catch (InvalidOperationException ex) when (ex.Message == "Can't read messages after the request is complete.")
+                {
+                    // RST_STREAM is processed by Kestrel before server reaches MoveNext.
+                }
+                finally
+                {
+                    tcs.SetResult(null);
+                }
+            }
+
+            // Arrange
+            var method = Fixture.DynamicGrpc.AddDuplexStreamingMethod<HelloRequest, HelloReply>(DuplexStreamingReadUntilEnd);
+
+            var channel = CreateChannel();
+
+            var client = TestClientFactory.Create(channel, method);
+
+            var lastLogCount = 0;
+            for (var i = 0; i < iterations; i++)
+            {
+                Logger.LogInformation($"Iteration {i}");
+
+                tcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                // Act
+                var call = client.DuplexStreamingCall();
+
+                await call.RequestStream.WriteAsync(new HelloRequest { Name = i.ToString() }).DefaultTimeout();
+                await call.ResponseStream.MoveNext().DefaultTimeout();
+
+                // No DefaultTimeout on CompleteAsync because we want dispose to happen as close as possible afterwards.
+                await call.RequestStream.CompleteAsync();
+                call.Dispose();
+
+                // Wait for server method to start exiting
+                await tcs.Task.DefaultTimeout();
+
+                // Assert
+
+                // Only test the logs for the last iteration.
+                var lastLogs = Logs.Skip(lastLogCount).ToList();
+                var sentEndStreamCount = lastLogs.Count(l =>
+                {
+                    // This is HTTP/2 specific. If this test is run with HTTP/3 then this check will need to be updated.
+                    if (l.EventId.Name == "Http2FrameReceived" &&
+                        l.State is IEnumerable<KeyValuePair<string, object>> state)
+                    {
+                        var frameType = state.Single(s => s.Key == "type").Value.ToString();
+                        var flags = state.Single(s => s.Key == "flags").Value.ToString();
+
+                        if (frameType == "DATA" && flags == "END_STREAM")
+                        {
+                            return true;
+                        }
+                    }
+
+                    return false;
+                });
```

Comment:
This is the only way I could figure out how to unit test this. I think if this test becomes troublesome then it could be deleted.